exporters/otlp/common/src/main/java/io/opentelemetry/exporter/otlp/internal/MetricsRequestMarshaler.java

```diff
@@ -0,0 +1,892 @@
+import java.io.IOException;
+import java.util.Collection;
+import java.util.List;
+import java.util.Map;
+
+/**
+ * {@link Marshaler} to convert SDK {@link MetricData} to OTLP ExportMetricsServiceRequest.
+ *
+ * <p>This class is internal and is hence not for public use. Its APIs are unstable and can change
+ * at any time.
+ */
+public final class MetricsRequestMarshaler extends MarshalerWithSize implements Marshaler {
+
+  private final ResourceMetricsMarshaler[] resourceMetricsMarshalers;
+
+  /**
+   * Returns a {@link MetricsRequestMarshaler} that can be used to convert the provided {@link
+   * MetricData} into a serialized OTLP ExportMetricsServiceRequest.
+   */
+  public static MetricsRequestMarshaler create(Collection<MetricData> metricDataList) {
+    Map<Resource, Map<InstrumentationLibraryInfo, List<Marshaler>>> resourceAndLibraryMap =
+        groupByResourceAndLibrary(metricDataList);
+
+    ResourceMetricsMarshaler[] resourceMetricsMarshalers =
+        new ResourceMetricsMarshaler[resourceAndLibraryMap.size()];
+    int posResource = 0;
+    for (Map.Entry<Resource, Map<InstrumentationLibraryInfo, List<Marshaler>>> entry :
+        resourceAndLibraryMap.entrySet()) {
+      final InstrumentationLibraryMetricsMarshaler[] instrumentationLibrarySpansMarshalers =
+          new InstrumentationLibraryMetricsMarshaler[entry.getValue().size()];
+      int posInstrumentation = 0;
+      for (Map.Entry<InstrumentationLibraryInfo, List<Marshaler>> entryIs :
+          entry.getValue().entrySet()) {
+        instrumentationLibrarySpansMarshalers[posInstrumentation++] =
+            new InstrumentationLibraryMetricsMarshaler(
+                InstrumentationLibraryMarshaler.create(entryIs.getKey()),
+                MarshalerUtil.toBytes(entryIs.getKey().getSchemaUrl()),
+                entryIs.getValue());
+      }
+      resourceMetricsMarshalers[posResource++] =
+          new ResourceMetricsMarshaler(
+              ResourceMarshaler.create(entry.getKey()),
+              MarshalerUtil.toBytes(entry.getKey().getSchemaUrl()),
+              instrumentationLibrarySpansMarshalers);
+    }
+
+    return new MetricsRequestMarshaler(resourceMetricsMarshalers);
+  }
+
+  private MetricsRequestMarshaler(ResourceMetricsMarshaler[] resourceMetricsMarshalers) {
+    super(calculateSize(resourceMetricsMarshalers));
+    this.resourceMetricsMarshalers = resourceMetricsMarshalers;
+  }
+
+  @Override
+  public void writeTo(CodedOutputStream output) throws IOException {
+    MarshalerUtil.marshalRepeatedMessage(
+        ExportMetricsServiceRequest.RESOURCE_METRICS_FIELD_NUMBER,
+        resourceMetricsMarshalers,
+        output);
+  }
+
+  private static int calculateSize(ResourceMetricsMarshaler[] resourceMetricsMarshalers) {
+    int size = 0;
+    size +=
+        MarshalerUtil.sizeRepeatedMessage(
+            ExportMetricsServiceRequest.RESOURCE_METRICS_FIELD_NUMBER, resourceMetricsMarshalers);
+    return size;
```

Comment:
? 
Some reason not to just return the value?

Reply:
I guess we're not expecting it, but I think we're using this pattern (I'm just following our current pattern) to make it easier to add / remove fields.

Reply:
hm. interesting. i'm surprised IDEA doesn't yell about this. :)